content/blog/2025-05-13-abstract-testing-floating-point.md

```diff
@@ -0,0 +1,201 @@
++++
+title = "What is a bug? New ways to test verification tools."
+[extra]
+latex = true
+bio = "Max Fan is a first-year CS PhD student at Cornell University."
+[[extra.authors]]
+name = "Max Fan"
+link = "https://max.fan"
++++
+
+My project is driven by a deceptively simple question: *What is a bug in a
+program verification tool?* 
+
+In the following blog post, I will:
+1. Explore a relaxed definition of a bug that ought to enable more effective
+   testing of verification tools. 
+2. Outline how to apply my definition to _abstractly test_ floating-point
+   verification software.
+3. Discuss my prototype implementation and preliminary results.
+
+```
+                                _______
+                               /       \
+                            ---|       |---
+                               |  Bug  |
+                            ---|       |---
+                               \_______/
+                                /  |  \
+                               /   |   \
+                              "    |    "
+```
+
+
+## What is a bug?
+
+Consider a verification tool **VERIFY** to take some program *p(x)* and provide
+a **GUARANTEE** such that for all inputs *x*, **GUARANTEE(**_p(x)_**)** holds.
+At first blush, one might define a bug as follows: 
+
+> <ins>Definition</ins> (_bug_). A bug in **VERIFY** exists iff there exists
+an input _x_ such that **GUARANTEE(**_p(x)_**)** does not hold.
+
+This is a perfectly reasonable definition. If you can violate a guarantee, it
+seems like you have a bug.
+
+But deep down in our hearts, there is _secretly_ a stronger notion of
+correctness that we'd like the verification tool to respect. In particular, it
+would be really bad if _I_ published a paper claiming my tool does **X** and
+prove **X** to be a sound technique when in reality it does **Y**. That sounds
+like a bug!
```

Comment:
I've kinda lost track of what X and Y are here, and how they relate to VERIFY and GUARANTEE above. Is X like GUARANTEE', a different property of program executions that implies the original GUARANTEE? A tiny bit more intuition might help here.

Reply:
Ah `**X**` and `**Y**` are meant to be different metavars tracking different things. I had thought that different casing and formatting would make this clear, but clearly it didn't work. Sorry about the confusion. I updated the text to say:
```diff
-would be really bad if _I_ published a paper claiming my tool does **X** and
-prove **X** to be a sound technique when in reality it does **Y**. That sounds
-like a bug!
+would be really bad if _I_ published a paper claiming my tool does **\$BLAH**
+and prove **\$BLAH** to be a sound technique when in reality it does **\$BLEH**.
+That sounds like a bug!
```

Reply:
Sure… I guess I'm asking mostly about the overall point you're getting at here. How do X and Y relate to VERIFY and GUARANTEE? One of the pairs sounds like "the stated goal of the tool" and one is maybe "the way the tool is formalized" or something, but what is the relationship between these two? Surely they have _some_ kind of relationship!!